### PR TITLE
Halogen 5

### LINKED ISSUE #25: Update for Halogen 5
Halogen is on its [fourth release candidate](https://github.com/slamdata/purescript-halogen/releases/tag/v5.0.0-rc.4) for v5, which appears imminent once docs are updated. [Formless](https://github.com/thomashoneyman/purescript-halogen-formless), the only Halogen dependency of this project, is updated for Halogen 5. As far as I'm aware the only other missing dependency is `RemoteData`, which has conflicting versions for `Lens`.

This project should be updated for Halogen 5 as soon as possible, where it can serve as a migration guide for others upgrading to Halogen 5 in the near future.

### PR DESCRIPTION
**Pull request by thomashoneyman**
Closes #25 by updating the project for Halogen 5. This update avoids updating any functionality in favor of a simple, direct translation from Halogen 4 to Halogen 5. There was also a major update in the form library, [Formless](https://github.com/thomashoneyman/purescript-halogen-formless), which required updating most page components more extensively than might be usual for a migration like this.

For more information on what's changed in Halogen 5, please see the [Changes in Halogen 5](https://github.com/slamdata/purescript-halogen/blob/master/docs/Changes%20in%20v5.md) documentation.

This was a fairly quick & pleasant transition, completed in about 2-3 hours in total.

### CODE REVIEW

File: src/Component/Router.purs [hunk 1]
```diff
@@ -79,40 +64,37 @@ component
...
-    when (route /= dest) do
-      H.modify_ _ { route = dest }
-    pure a
-
-  render :: State -> H.ParentHTML Query ChildQuery ChildSlot m
+  render :: State -> H.ComponentHTML Void ChildSlots m
   render { route } = case route of
     Home -> 
-      HH.slot' CP.cp1 unit Home.component unit absurd
+      HH.slot (SProxy :: _ "home") unit Home.component unit absurd
```
Comment (thomashoneyman) [inline src/Component/Router.purs]: In Halogen 4 we used a child path and slot type to identify a particular component among many. In Halogen 5 we use a symbol proxy and a slot type instead. 

A symbol proxy is a type-level string. You've seen them before: they're used as labels for row types (and records and variants, which are built on top of rows). In this case we're identifying the component with the label `home` in our `ChildSlots` row defined at the top of this module.

I've written `SProxy :: _ "home"` as a shorthand here, but the full type is `SProxy :: SProxy "home"`. That means that at the value level we have just `SProxy`, but this value has the type `SProxy "home"`, which is how we're able to use the label "home" at the type level.

File: src/Component/Router.purs [hunk 2]
```diff
@@ -43,29 +42,15 @@ data Query a
...
-
-type ChildSlot = Either7
-  Unit
-  Unit
-  Unit
-  Unit
-  Unit
-  Unit
-  Unit
+type ChildSlots =
```
Comment (thomashoneyman) [inline src/Component/Router.purs:54]: In Halogen 5 we use a row type to describe child components, instead of the old nested either / coproduct pairing. It's quite a bit cleaner and allows you to easily give descriptive names to each component.

A child path used to be a pairing of a component's query type (what a parent can use to trigger behaviors in or request information from one of its child components) and slot type (a unique address when there are multiple components of the same type). 

In Halogen 5, we instead have a dedicated `Slot` type which uses the child component's query type, message type (what outputs are possible to handle from the component), and slot address.

For example:

```purescript
-- this parent component has one child component type, labeled `component`, with the query
-- type `ComponentQuery`, outputs of type `ComponentMessage`, and we'll distinguish among
-- multiple copies of the component using an `Int` address.
type MyChildSlots =
  ( component :: H.Slot ComponentQuery ComponentMessage Int )
```

Our router deals with page components which raise no output messages and have no public query interface you can use to request information or manipulate them. They simply aren't necessary. So for each of our page components the slot type is just:

```purescript
type ChildSlots =
  ( home :: H.Slot (Const Void) Void Unit )
```

In other words, `H.Slot` expects a query type of kind `Type -> Type`, a message type of kind `Type`, and a slot address of kind `Type`. When you need to provide a type but have no possible values that could satisfy the type, you can use `Void` -- a type with no accompanying values. And when you need to provide something of kind `Type -> Type` but have no possible values, you can use `Const Void`. So when we have a component with no queries and no messages we can use `Const Void` and `Void` to represent that.

This is fairly common in our application, so I've defined a helper type in `Component.Utils` for components with no queries or messages:

```purescript
type OpaqueSlot = H.Slot (Const Void) Void
```

File: src/Page/Editor.purs [hunk 3]
```diff
@@ -48,32 +48,31 @@ type State =
 type Input =
   { slug :: Maybe Slug }
 
-type ChildQuery m = F.Query Query TagInput.Query Unit EditorFields m
+type ChildSlots =
+  ( formless :: F.Slot EditorFields (Const Void) FormChildSlots Article Unit )
```
Comment (thomashoneyman) [inline src/Page/Editor.purs:55]: Formless is the component that has undergone the greatest number of changes as part of the Halogen 5 transition. For the full details, [see the Formless library](https://github.com/thomashoneyman/purescript-halogen-formless). If you have written a form and you want to extend the form's state, query algebra, actions, messages, child components, or anything else, that information no longer leaks out to the parent component but instead is handled entirely within Formless. Accordingly, any extensions you want to make to the component will show up in its slot type.

The main thing that affects us here is that the Formless slot type now takes:

- Your form type being run in Formless (here, `EditorFields`)
- Any queries you wish to extend the component with (here, none, represented with `Const Void`
- Any child components the form will have (here, `FormChildSlots`, defined at the end of the module)
- Any messages the component will output (here, we'll get an `Article` on successful submission)
- The form slot address

File: src/Page/Editor.purs [hunk 4]
```diff
@@ -146,57 +132,84 @@ newtype EditorFields r f = EditorFields (r
...
-validators = EditorFields
-  { title: V.required >>> V.minLength 1
-  , description: V.required >>> V.minLength 1
-  , body: V.required >>> V.minLength 3
-  , tagList: F.hoistFn_ (map unwrap)
+type FormChildSlots = 
+  ( tagInput :: H.Slot (Const Void) TagInput.Message Unit )
+
+data FormAction 
+  = HandleTagInput TagInput.Message
```
Comment (thomashoneyman) [inline src/Page/Editor.purs:140]: Formless can now be extended with more actions as you see fit. Here, we'll extend the component with the ability to handle messages being output by the `TagInput` component. Notice how the parent component, the `Editor` page, doesn't have to care about the tag input anymore because it's encapsulated within the form component.

File: src/Page/Editor.purs [hunk 5]
```diff
@@ -146,57 +132,84 @@ newtype EditorFields r f = EditorFields (r
...
+
+formInput :: forall m. Monad m => F.Input' EditorFields m
+formInput = 
+  { validators: EditorFields
+      { title: V.required >>> V.minLength 1
+      , description: V.required >>> V.minLength 1
+      , body: V.required >>> V.minLength 3
+      , tagList: F.hoistFn_ (map unwrap)
+      }
+  , initialInputs: Nothing
```
Comment (thomashoneyman) [inline src/Page/Editor.purs]: Formless no longer takes a render function as part of its input. Instead, the render function is given as part of its `spec` (below). In addition, it's no longer necessary to provide a set of starting values for the fields in your form -- if they have a valid `Initial` instance (almost all types in a usual form do) then Formless will generate these values for you.

File: src/Page/Editor.purs [hunk 6]
```diff
@@ -146,57 +132,84 @@ newtype EditorFields r f = EditorFields (r
...
-  HH.form_
-    [ HH.fieldset_
-      [ title
-      , description
-      , body
-      , HH.slot unit TagInput.component unit (HE.input $ F.raise <<< H.action <<< HandleTagInput)
-      , Field.submit $ if isJust mbArticle then "Commit changes" else "Publish"
-      ]
-    ]
+  -> F.Spec EditorFields () (Const Void) FormAction FormChildSlots Article m
```
Comment (thomashoneyman) [inline src/Page/Editor.purs]: The Formless component now takes a `Spec`, which describes all the ways in which you'd like to override or extend the component. At minimum you will want to provide your own render function to render out your form fields. That could be done with just:

```
formSpec = F.defaultSpec { render  = render }
  where
  render { form } = ...
```

However, you can extend a form component in many more ways, represented in the `Spec` type above. In order, you should provide:

- the form type the form is running (here, `EditorFields`)
- any extra fields you want present in `State` (here, none, represented by `()`)
- any additional queries you want the component to be able to run (here, none, represented by `(Const Void)`)
- any additional actions you want the component to be able to run (here, `FormAction`, which handles messages from the `TagInput` component)
- any child components you plan to mount in the render function in the component (here, `FormChildSlots`, which describes the `TagInput` component)
- what messages the component will output, which you can define (here, a valid `Article` will be produced on successful form submission)
- what monad the component will run in (here, any monad supporting `MonadAff` -- including `AppM`)

If you choose to extend the component's query or action types, then you'll also want to write handlers for them: what should the component do when your query or action is triggered? That can be done by providing a `handleAction` or `handleQuery` function in the spec alongside your provided render function. 

Below, we've also intercepted the Formless `Submitted` message with `handleMessage`. There are more ways you can extend the spec, like hooking in to the `receive`, `initialize`, and `finalize` events; if you need that advanced behavior, please see the Formless library.

Most simple forms can just override the `render` function and possibly provide their own messages with `handleMessage`.

**Comment by thomashoneyman** [comment]
Due to the long delay on the Halogen 5 release, I'm leaning towards merging this into master this week. However, if any watchers of this repository have a reason why I shouldn't do this, please feel free to comment!

**Comment by JordanMartinez** [comment]
I'd say just go ahead and merge it. I don't know when Gary will make the next release, even though it's quite useable as is.

**Comment by willisplummer** [comment]
It makes sense to me to merge -- I wanted to use Halogen 5 and having this as an example was useful for understanding what was different. Maybe add a link to this PR in the README or even tag the repo before merging so that it's easy for ppl to see what it looked like at 4 vs. 5?

**Comment by thomashoneyman** [comment]
That's a good idea -- I think a tag + a link in the readme would help make sure people on Halogen 4 and Halogen 5 are both able to get value from the repo.
